lib/tools/profiler.js

```diff
@@ -0,0 +1,118 @@
+'use strict';
+
+const v8Profiler = require('v8-profiler');
+const EventEmitter = require('events');
+
+// 30 min
+const DEFAULT_INTERVAL_MS = 1800000;
+const NAME_PREFIX = 'v8.heap.snapshot';
+
+/**
+ * The Profiler
+ */
+class Profiler extends EventEmitter {
+
+  /**
+   * Create a new Profiler.
+   *
+   * @params {Object} options - Object with the profiler configuration options
+   *                  interval: The amount in milliseconds to wait before taking snapshots
+   *                  tag: custom identifier
+   *                  extension: The extension of the file. It must be of the type:
+   *                             .cpuprofile, .heapsnapshot, .heapprofile, or .heaptimeline
+   */
+  constructor(options) {
+    super();
+    this.extension = this._sanitize(options.extension) || '';
+    this.tag = this._sanitize(options.tag) || '';
+    this.interval = parseInt(options.interval, 10) || DEFAULT_INTERVAL_MS;
+    this.namePrefix = this._sanitize(options.namePrefix) || NAME_PREFIX;
+    this.intervalId = null;
+    this.on('takeSnapshot', this.takeSnapshot);
+  }
+
+  /**
+   * Emits the intervalStarted event and calls the startInterval func
+   */
+  start() {
+    this.startInterval();
+    this.emit('intervalStarted');
+  }
+
+  /**
+   * Emits the intervalStopped event and calls the stopInterval func
+   */
+  stop() {
+    this.stopInterval();
+    this.emit('intervalStopped');
+  }
+
+  /**
+   * Stores the value of the intervalId for later recycling
+   *
+   * @return {Number} - interval id
+   */
+  startInterval() {
+    this.intervalId = setInterval(() => {
+      this.emit('takeSnapshot');
+    }, this.interval);
+    return this.intervalId;
+  }
+
+  /**
+   * clears the Interval
+   */
+  stopInterval() {
+    clearInterval(this.intervalId);
+  }
+
+  /**
+   * Emits the snapshotCreated event and sends the captured snapshot as an argument
+   */
+  takeSnapshot() {
+    const name = this._getName();
+    this.emit('snapshotCreated', v8Profiler.takeSnapshot(name));
+  }
+
+  /**
+   * Utility Functions
+   */
+
+  /**
+   * Gets the epoch timestamp
+   *
+   * @return {Number} - the number of milliseconds since 1970/01/01
+   */
+  _getTimestamp() {
+    const date = new Date();
+    return date.getTime();
+  }
+
+  /**
+  * It generates the name of the snapshot.
+  * The snapshot's name follows the next pattern:
+  *    <PREFIX>_<TAG>_<EPOCH TIMESTAMP>
+  *    Example: v8.heap.snapshot_iamatag_1491242331638
+  *
+  * @param {Number} - optional timestamp
+  *
+  * @returns {String}
+  */
+  _getName(timestamp) {
```

Comment:
No need to prefix functions with `_`: https://github.com/airbnb/javascript#naming--leading-underscore